pkg/ccl/changefeedccl/helpers_test.go

```diff
@@ -676,6 +580,226 @@ func closeFeed(t testing.TB, f cdctest.TestFeed) {
+) (testServer TestServer, cleanup func()) {
+	options := makeOptions(opts...)
+	return makeTenantServerWithOptions(t, options)
+}
+func makeTenantServerWithOptions(
+	t *testing.T, options feedTestOptions,
+) (testServer TestServer, cleanup func()) {
+	systemServer, systemDB, clusterCleanup := startTestFullServer(t, options)
+	tenantID, tenantServer, tenantDB, tenantCleanup := startTestTenant(t, systemServer, options)
+
+	return TestServer{
+			DB:           tenantDB,
+			Server:       tenantServer,
+			SystemDB:     systemDB,
+			SystemServer: systemServer,
+			TestingKnobs: tenantServer.(*server.TestTenant).Cfg.TestingKnobs,
+			Codec:        keys.MakeSQLCodec(tenantID),
+		}, func() {
+			tenantCleanup()
+			clusterCleanup()
+		}
+}
+
+func makeServer(t *testing.T, opts ...feedTestOption) (testServer TestServer, cleanup func()) {
+	options := makeOptions(opts...)
+	return makeServerWithOptions(t, options)
+}
+
+func makeServerWithOptions(
+	t *testing.T, options feedTestOptions,
+) (testServer TestServer, cleanup func()) {
+	if options.useTenant {
+		t.Logf("making server as secondary tenant")
+		return makeTenantServerWithOptions(t, options)
+	}
+	t.Logf("making server as system tenant")
+	return makeSystemServerWithOptions(t, options)
+}
+
+func randomSinkType(opts ...feedTestOption) string {
+	options := makeOptions(opts...)
+	return randomSinkTypeWithOptions(options)
+}
+
+func randomSinkTypeWithOptions(options feedTestOptions) string {
+	sinkWeights := map[string]int{
+		"kafka":        2, // run kafka a bit more often
+		"enterprise":   1,
+		"webhook":      1,
+		"pubsub":       1,
+		"sinkless":     0, // default to no sinkless to avoid surprises
+		"cloudstorage": 0, // requires externalIODir set
+	}
+	if options.includeSinkless {
+		sinkWeights["sinkless"] = 1
+	}
+	if options.externalIODir != "" {
+		sinkWeights["cloudstorage"] = 1
+	}
+	if options.allowedSinkTypes != nil {
+		sinkWeights = map[string]int{}
+		for _, sinkType := range options.allowedSinkTypes {
+			sinkWeights[sinkType] = 1
+		}
+	}
+	if options.disabledSinkTypes != nil {
+		for _, sinkType := range options.disabledSinkTypes {
+			sinkWeights[sinkType] = 0
+		}
+	}
+	weightTotal := 0
+	for _, weight := range sinkWeights {
+		weightTotal += weight
+	}
+	p := rand.Float32() * float32(weightTotal)
+	var sum float32 = 0
+	for sink, weight := range sinkWeights {
+		sum += float32(weight)
+		if p <= sum {
+			return sink
+		}
+	}
+	return "kafka" // unreachable
+}
+
+// addCloudStorageOptions adds the options necessary to enable a server to run a
+// cloudstorage changefeed on it
+func addCloudStorageOptions(t *testing.T, options *feedTestOptions) (cleanup func()) {
+	dir, dirCleanupFn := testutils.TempDir(t)
+	options.externalIODir = dir
+	oldKnobsFn := options.knobsFn
+	options.knobsFn = func(knobs *base.TestingKnobs) {
+		if oldKnobsFn != nil {
+			oldKnobsFn(knobs)
+		}
+		blobClientFactory := blobs.NewLocalOnlyBlobClientFactory(options.externalIODir)
```

Comment:
if this is only needed for nodelocal:// -- i.d just not use nodelocal :)

Reply:
As opposed to a different BlobClientFactory?

Reply:
Basically, I just want to have fewest possible number of special cases.  I was just wondering if the fact
that you need to create this blob client factory is driven because we use nodelocal:// in some tests?

Reply:
Unfortunately it was required for a bunch of tests to work even if they didn't mention nodelocal